tools/PI/DevHome.PI/Controls/SystemBar.cs

```diff
@@ -0,0 +1,150 @@
+// one which implements a system bar in the horizontal position, another in a vertical position
+public partial class SystemBar : UserControl, INotifyPropertyChanged
+{
+    private bool _isSnapped;
+
+    // This is used to toggle the "snap to an app" icon to have different values
+    // based on if we are snapped or not
+    public bool IsSnapped
+    {
+        get => _isSnapped;
+        set
+        {
+            _isSnapped = value;
+            CurrentSnapButtonText = _isSnapped ? UnsnapButtonText : SnapButtonText;
+        }
+    }
+
+    private bool _isSnappingEnabled;
+
+    // This is used to enable the "snap to an app" icon if we're allowed to snap
+    public bool IsSnappingEnabled
+    {
+        get => _isSnappingEnabled;
+        set
+        {
+            _isSnappingEnabled = value;
+            OnPropertyChanged(nameof(IsSnappingEnabled));
+        }
+    }
+
+    private bool _isMaximized;
+
+    // This is used determine if the system bar should treat the window as maximized or not. That
+    // allows us to enable/disable the "Maximize" button and the "Restore" button in the titlebar
+    public bool IsMaximized
+    {
+        get => _isMaximized;
+        set
+        {
+            _isMaximized = value;
+            MaximizeButtonVisibility = _isMaximized ? Visibility.Collapsed : Visibility.Visible;
+            RestoreButtonVisibility = _isMaximized ? Visibility.Visible : Visibility.Collapsed;
+            OnPropertyChanged(nameof(MaximizeButtonVisibility));
+            OnPropertyChanged(nameof(RestoreButtonVisibility));
+        }
+    }
+
+    protected Visibility MaximizeButtonVisibility { get; private set; } = Visibility.Visible;
+
+    protected Visibility RestoreButtonVisibility { get; private set; } = Visibility.Collapsed;
+
+    private const string UnsnapButtonText = "\ue89f";
+    private const string SnapButtonText = "\ue8a0";
+
+    private string _currentSnapButtonText = SnapButtonText;
+
+    protected string CurrentSnapButtonText
+    {
+        get => _currentSnapButtonText;
+        private set
+        {
+            if (_currentSnapButtonText != value)
+            {
+                _currentSnapButtonText = value;
+                OnPropertyChanged(nameof(CurrentSnapButtonText));
+            }
+        }
+    }
+
+    public SystemBar()
+    {
+    }
+
+    public void Initialize()
+    {
+        IsSnappingEnabled = TargetAppData.Instance.HWnd != HWND.Null;
+        TargetAppData.Instance.PropertyChanged += TargetApp_PropertyChanged;
+    }
+
+    private void TargetApp_PropertyChanged(object? sender, PropertyChangedEventArgs e)
+    {
+        if (e.PropertyName == nameof(TargetAppData.HWnd))
+        {
+            IsSnappingEnabled = TargetAppData.Instance.HWnd != HWND.Null;
+        }
+    }
+
+    protected void SnapButton_Click(object sender, RoutedEventArgs e)
+    {
+        var barWindow = Application.Current.GetService<PrimaryWindow>().DBarWindow;
```

Comment:
nit: should consider plumbing through a call to GetRequiredService if we always expect this to be non-null.

Reply:
It doesn't seem like GetRequiredService is a pattern in Dev Home, maybe something we should consider converting to in the future.